src/ipc/capnp/common-types.h

```diff
@@ -256,6 +256,146 @@ decltype(auto) CustomReadField(TypeList<util::Result<LocalType>>, Priority<1>, I
+                                                          std::declval<LocalType&>()))* enable = nullptr)
+{
+    return read_dest.update([&](auto& value) { CustomReadMessage(invoke_context, reader.get(), value); });
+}
+
+//! Helper for CustomPassField below. Call Accessor::get method if it has one,
+//! otherwise return capnp::Void.
+template <typename Accessor, typename Message>
+decltype(auto) MaybeGet(Message&& message, decltype(Accessor::get(message))* enable = nullptr)
+{
+    return Accessor::get(message);
+}
+
+template <typename Accessor>
+::capnp::Void MaybeGet(...)
+{
+    return {};
+}
+
+//! Helper for CustomPassField below. Call Accessor::init method if it has one,
+//! otherwise do nothing.
+template <typename Accessor, typename Message>
+decltype(auto) MaybeInit(Message&& message, decltype(Accessor::get(message))* enable = nullptr)
+{
+    return Accessor::init(message);
+}
+
+template <typename Accessor>
+::capnp::Void MaybeInit(...)
+{
+    return {};
+}
+
+//! Overload CustomPassField to serialize types of objects that have
+//! CustomPassMessage overloads. Defining PassMessage overloads is simpler than
+//! defining PassField overloads, because PassMessage overloads can be normal
+//! functions instead of template functions, and defined in a .cpp file instead
+//! of a header.
+//!
+//! Defining a PassField or PassMessage overload is useful for input/output
+//! parameters. If an overload is not defined these parameters will just be
+//! deserialized on the server side with ReadField into a temporary variable,
+//! then the server method will be called passing the temporary variable as a
+//! parameter, then the temporary variable will be serialized and sent back to
+//! the client with BuildField. But if a PassField or PassMessage overload is
+//! defined, the overload is called with a callback to invoke and pass
+//! parameters to the server side function, and run arbitrary code before and
+//! after invoking the function.
+template <typename Accessor, typename... LocalTypes, typename ServerContext, typename Fn, typename... Args>
+auto CustomPassField(TypeList<LocalTypes...>, ServerContext& server_context, Fn&& fn, Args&&... args)
+    -> decltype(CustomPassMessage(server_context, MaybeGet<Accessor>(server_context.call_context.getParams()),
+                                  MaybeGet<Accessor>(server_context.call_context.getResults()), nullptr))
+{
+    CustomPassMessage(server_context, MaybeGet<Accessor>(server_context.call_context.getParams()),
+                      MaybeInit<Accessor>(server_context.call_context.getResults()),
+                      [&](LocalTypes... param) { fn.invoke(server_context, std::forward<Args>(args)..., param...); });
+}
+
+//! Generic ::capnp::Data field builder for any class that a Span can be
+//! constructed from, particularly BaseHash and base_blob classes and
+//! subclasses. It's also used to serialize vector<unsigned char>
+//! GCSFilter::ElementSet set elements.
```

Comment:
Should this say ..."vector\<unsigned char\> *and* GCSFilter::ElementSet set elements."?